Chatkit/Subscriptions/SubscriptionManager.swift

```diff
@@ -0,0 +1,79 @@
+
+protocol HasSubscriptionManager {
+    var subscriptionManager: SubscriptionManager { get }
+}
+
+protocol SubscriptionManager {
+    func subscribe(toType subscriptionType: SubscriptionType, sender: AnyObject, completion: @escaping SubscribeHandler)
+    func unsubscribe(fromType subscriptionType: SubscriptionType, sender: AnyObject)
+    func unsubscribeFromAll()
+}
+
+class ConcreteSubscriptionManager: SubscriptionManager {
+    
+    typealias Dependencies = HasSubscriptionFactory
+    typealias SubscriptionDetails = (subscription: Subscription, subscribers: NSHashTable<AnyObject>)
+    
+    private let dependencies: Dependencies
+    private var subscriptionDetailsByType: [SubscriptionType: SubscriptionDetails] = [:]
+    
+    init(dependencies: Dependencies) {
+        self.dependencies = dependencies
+    }
+    
+    // MARK: SubscriptionManager
+    
+    func subscribe(toType subscriptionType: SubscriptionType, sender: AnyObject, completion: @escaping SubscribeHandler) {
+        
+        let subscriptionDetails: SubscriptionDetails
+        
+        // Check if we are already have SubscriptionDetails for the specified SubscriptionType and if so use that
+        if let subscriptionDetailsTmp = subscriptionDetailsByType[subscriptionType] {
+            subscriptionDetails = subscriptionDetailsTmp
+        } else {
+            let subscription = dependencies.subscriptionFactory.makeSubscription(subscriptionType: subscriptionType)
+            let subscribers = NSHashTable<AnyObject>.weakObjects()
+            subscriptionDetails = (subscription: subscription, subscribers: subscribers)
+            subscriptionDetailsByType[subscriptionType] = subscriptionDetails
+        }
+
+        // Hold a *weak* reference to the sender as a "subscriber"
+        subscriptionDetails.subscribers.add(sender)
+        
+        // TODO: Subscription unhappy paths
+        // It was decided to defer work on handling Subscription *un*happy paths in favour
+        // of shipping the happy paths on the SDK
+        // Do we need to remove the subscription from subscriptionsByType if the subscribe call fails?
+        // At the moment I am thinking no because if someone calls subscribe again it should cause the dead
+        // subscription to come back to life and reattempt connection
+        
+        subscriptionDetails.subscription.subscribe(completion: completion)
```

Comment:
Wouldn't it be safer to subscribe only when we create a new subscription?

Reply:
This behaviour effectively happens within the `Subscription` itself (if `subscribe` is called twice then it effectvely ignore's the second call)  Its therefore unncessary to reimplement this logic here and doing so would add extra complexity IMO.  I also think the `SubscriptionManager` should be ignorant of the `Subscription`s state and not ever assume that it is connected.